commons/src/main/scala/com/lynbrookrobotics/potassium/commons/drivetrain/PurePursuitControllers.scala

```diff
@@ -0,0 +1,159 @@
+      }
+    )
+  }
+
+  def headingToPoint(start: Point, end: Point): Angle = {
+      val diff = end - start
+      Radians(Math.atan2(
+        diff.y.toFeet,
+        diff.x.toFeet
+      ))
+  }
+
+  def trigonemtricToCompass(angle: Angle): Angle = {
+    Degrees(90) - angle
+  }
+
+  /**
+    * see: https://www.mathworks.com/help/robotics/ug/pure-pursuit-controller.html
+    * @param angle the current compass angle, with the0 degrees defined at the y axis.
+    * @param position the current x,y,z position of the robot
+    * @param biSegmentPath the path to follow
+    * @return turn output, what to multiply forward output by (1 for forward, -1 for backwards),
+    *         lookAheadPoint
+    */
+  def purePursuitControllerTurn(angle: Signal[Angle],
+                                position: PeriodicSignal[Point],
+                                biSegmentPath: Signal[(Segment, Option[Segment])])
+                               (implicit properties: Signal[UnicycleProperties]): (PeriodicSignal[Dimensionless],
+                                                                                   PeriodicSignal[Double],
+                                                                                   PeriodicSignal[Point]) = {
+    val lookAheadPoint = position.zip(properties).zip(biSegmentPath).map{p =>
+      val ((pose, props), path) = p
+      getLookAheadPoint(
+        path,
+        pose,
+        props.defaultLookAheadDistance
+      )
+    }
+
+    val heading = position.zip(lookAheadPoint).map(
+      p => headingToPoint(p._1, p._2)).map(trigonemtricToCompass(_))
+
+    val clampedTarget = heading.zip(angle).map { t =>
+      val (heading, angle) = t
+      val error = heading - angle
+
+      if (error <= Degrees(-90)) {
+//        println(s"going back, angle error ${angle.toDegrees}")
+        (heading + Degrees(180), -1D)
+      } else if (error >= Degrees(90)) {
+//        println(s"going back, angle error ${angle.toDegrees}")
+        (heading - Degrees(180), -1D)
+      } else {
+        (heading, 1D)
+      }
+    }
+
+    (
+      PIDF.pid(
+        angle.toPeriodic,
+        clampedTarget.map(_._1),
+        properties.map(_.turnPositionControlGains)),
+      clampedTarget.map(_._2),
+      lookAheadPoint
+    )
+  }
+
+
+
+  @deprecated
+  def getLookExtrapolatedAheadPoint(biSegmentPath: (Segment, Option[Segment]),
+    currPosition: Point,
+    lookAheadDistance: Length): Point = {
+    import MathUtilities._
+    val firstLookAheadPoint = intersectionLineCircleFurthestFromStart(
+      biSegmentPath._1,
+      currPosition,
+      lookAheadDistance)
+
+    val secondLookAheadPoint = biSegmentPath._2.flatMap { s =>
+      intersectionClosestToEnd(s, currPosition, lookAheadDistance)
+    }
+
+    secondLookAheadPoint.getOrElse(
+      firstLookAheadPoint.getOrElse {
+        println(s"expanding look ahead distance to ${1.1 * lookAheadDistance}")
```

Comment:
I left this here because if this happens, our constants are waaaaaaay off and need to be seriously tuned